### ISSUE TITLE
Fails to ignore lines with Ethiopian (utf-8) characters

### ISSUE THREAD

**Issue by skangas** [codespell-project/codespell#2604]
I have a file with these lines, containing Ethiopian (utf-8) characters:

```
("`see" "ሤ") ("ssee" "ሤ") ("s2ee" "ሤ")
("wE" ?ዌ) ("wee" "ዌ")
("`See" "ፄ") ("SSee" "ፄ") ("S2ee" "ፄ")
```

When I add these lines to `foo.txt` and pass codespell the `-x foo.txt` flag, codespell fails to ignore the corresponding lines.

**Comment by DimitriPapadopoulos** [comment]
Why should it ignore these lines?

**Comment by skangas** [comment]
> Why should it ignore these lines?

When I do _not_ put these lines in the ignore file using `-x`, codespell gives me the below diff. I think all of these are mistakes, and should not be corrected, hence I want to ignore them.

```
diff --git a/lisp/language/ethio-util.el b/lisp/language/ethio-util.el
index 2f76acfe7c..48528787d5 100644
--- a/lisp/language/ethio-util.el
+++ b/lisp/language/ethio-util.el
@@ -1123,7 +1123,7 @@ "ethiopic-sera"
  ("`si" ?ሢ) ("ssi" "ሢ") ("s2i" "ሢ")
  ("`sa" ?ሣ) ("ssa" "ሣ") ("s2a" "ሣ")
  ("`sE" ?ሤ) ("ssE" "ሤ") ("s2E" "ሤ")
-   ("`see" "ሤ") ("ssee" "ሤ") ("s2ee" "ሤ")
+   ("`see" "ሤ") ("see" "ሤ") ("s2ee" "ሤ")
  ("`s" ?ሥ) ("ss" "ሥ") ("s2" "ሥ")
  ("`so" ?ሦ) ("sso" "ሦ") ("s2o" "ሦ")
  ("`sWa" ?ሧ) ("ssWa" "ሧ") ("s2Wa" "ሧ")
@@ -1292,7 +1292,7 @@ "ethiopic-sera"
  ("wu" ?ዉ)
  ("wi" ?ዊ)
  ("wa" ?ዋ)
- ("wE" ?ዌ) ("wee" "ዌ")
+ ("wE" ?ዌ) ("we" "ዌ")
  ("w" ?ው)
  ("wo" ?ዎ)
 
@@ -1420,7 +1420,7 @@ "ethiopic-sera"
  ("`Si" ?ፂ) ("SSi" "ፂ") ("S2i" "ፂ")
  ("`Sa" ?ፃ) ("SSa" "ፃ") ("S2a" "ፃ")
  ("`SE" ?ፄ) ("SSE" "ፄ") ("S2E" "ፄ")
-   ("`See" "ፄ") ("SSee" "ፄ") ("S2ee" "ፄ")
+   ("`See" "ፄ") ("see" "ፄ") ("S2ee" "ፄ")
  ("`S" ?ፅ) ("SS" "ፅ") ("S2" "ፅ")
  ("`So" ?ፆ) ("SSo" "ፆ") ("S2o" "ፆ")
```

**Comment by DimitriPapadopoulos** [comment]
Other users do want lines that contain non-ASCII characters to be checked. For example, I want codespell to catch the misspelled `capalize` function name, although the default argument is non-ASCII:
```python
def capalize(letter="ፄ"):
   pass
```

So let me rephrase. What you actually want is to skip words written with ASCII characters but are not English words, such as `ssee`, `wee`, `SSee`.

Unfortunately, codespell cannot help here. There is no straightforward, universal way to infer whether some possible typos are actual typos, or domain-specific words or words from other languages. It is even worse with files that contain words from languages that can be written using the 26 letters of the Latin alphabet exclusively  – such as Malay, Indonesian, Dutch.

You will have to ignore such words explicitly using option `-L`.

**Comment by skangas** [comment]
> What you actually want is to ignore words written with ASCII characters but are not English words, such as `ssee`, `wee`, `SSee`.

No, this is not what I am asking for, not at all. I must have failed to communicate the issue clearly, so please allow me to try again:

I want lines to be ignored if and only if they are added to a file, which is given to codespell using the `-x` (`--exclude-file`) flag. That flag is documented to function like this:

```
       -x FILE, --exclude-file FILE
              ignore whole lines that match those in the file FILE.  The lines in FILE should
              match the to-be-excluded lines exactly
```

What is unexpected is that lines exactly matching lines in `FILE` are _not_ ignored in some cases. More specifically, I had issues with lines containing Ethiopian characters. Here is what I see:

```
$ # download the problematic file to /tmp/foo.txt
$ curl -s -o /tmp/foo.txt https://git.savannah.gnu.org/gitweb/\?p\=emacs.git\;a\=blob_plain\;f\=lisp/language/ethio-util.el\;hb\=HEAD
$ cp /tmp/foo.txt{,.orig}
$ codespell --write-changes --summary /tmp/foo.txt
WARNING: Decoding file using encoding=utf-8 failed: /tmp/foo.txt
WARNING: Trying next encoding iso-8859-1
/tmp/foo.txt:1188: bu ==> by, be, but, bug, bun, bud, buy, bum
/tmp/foo.txt:1188: Bu ==> By, Be, But, Bug, Bun, Bud, Buy, Bum
/tmp/foo.txt:1205: te ==> the, be, we, to
/tmp/foo.txt:1209: tE ==> the, be, we, to
/tmp/foo.txt:1390: Te ==> The, Be, We, To
/tmp/foo.txt:1394: TE ==> THE, BE, WE, TO
/tmp/foo.txt:1445: fo ==> of, for, to, do, go
/tmp/foo.txt:1445: Fo ==> Of, For, To, Do, Go
FIXED: /tmp/foo.txt

-------8<-------
SUMMARY:
ssee          2
wee           1
$ diff -u /tmp/foo.txt{,.orig}
--- /tmp/foo.txt	2022-11-17 17:03:32.883297798 +0100
+++ /tmp/foo.txt.orig	2022-11-17 17:03:26.323263279 +0100
@@ -1123,7 +1123,7 @@
  ("`si" ?ሢ) ("ssi" "ሢ") ("s2i" "ሢ")
  ("`sa" ?ሣ) ("ssa" "ሣ") ("s2a" "ሣ")
  ("`sE" ?ሤ) ("ssE" "ሤ") ("s2E" "ሤ")
-   ("`see" "ሤ") ("see" "ሤ") ("s2ee" "ሤ")
+   ("`see" "ሤ") ("ssee" "ሤ") ("s2ee" "ሤ")
  ("`s" ?ሥ) ("ss" "ሥ") ("s2" "ሥ")
  ("`so" ?ሦ) ("sso" "ሦ") ("s2o" "ሦ")
  ("`sWa" ?ሧ) ("ssWa" "ሧ") ("s2Wa" "ሧ")
@@ -1292,7 +1292,7 @@
  ("wu" ?ዉ)
  ("wi" ?ዊ)
  ("wa" ?ዋ)
- ("wE" ?ዌ) ("we" "ዌ")
+ ("wE" ?ዌ) ("wee" "ዌ")
  ("w" ?ው)
  ("wo" ?ዎ)
 
@@ -1420,7 +1420,7 @@
  ("`Si" ?ፂ) ("SSi" "ፂ") ("S2i" "ፂ")
  ("`Sa" ?ፃ) ("SSa" "ፃ") ("S2a" "ፃ")
  ("`SE" ?ፄ) ("SSE" "ፄ") ("S2E" "ፄ")
-   ("`See" "ፄ") ("see" "ፄ") ("S2ee" "ፄ")
+   ("`See" "ፄ") ("SSee" "ፄ") ("S2ee" "ፄ")
  ("`S" ?ፅ) ("SS" "ፅ") ("S2" "ፅ")
  ("`So" ?ፆ) ("SSo" "ፆ") ("S2o" "ፆ")
```

Now, I add to `/tmp/foo-ignore.txt` the following lines:

```
   ("`See" "ፄ") ("see" "ፄ") ("S2ee" "ፄ")
 ("wE" ?ዌ) ("we" "ዌ")
   ("`see" "ሤ") ("see" "ሤ") ("s2ee" "ሤ")
```

And I then rerun codespell like so (note the added `--exclude-file` flag):

```
$ # download the problematic file to /tmp/foo.txt
$ curl -s -o /tmp/foo.txt https://git.savannah.gnu.org/gitweb/\?p\=emacs.git\;a\=blob_plain\;f\=lisp/language/ethio-util.el\;hb\=HEAD
$ cp /tmp/foo.txt{,.orig}
$ codespell --write-changes --summary --exclude-file /tmp/foo-ignore.txt /tmp/foo.txt
WARNING: Decoding file using encoding=utf-8 failed: /tmp/foo.txt
WARNING: Trying next encoding iso-8859-1
/tmp/foo.txt:1188: bu ==> by, be, but, bug, bun, bud, buy, bum
/tmp/foo.txt:1188: Bu ==> By, Be, But, Bug, Bun, Bud, Buy, Bum
/tmp/foo.txt:1205: te ==> the, be, we, to
/tmp/foo.txt:1209: tE ==> the, be, we, to
/tmp/foo.txt:1390: Te ==> The, Be, We, To
/tmp/foo.txt:1394: TE ==> THE, BE, WE, TO
/tmp/foo.txt:1445: fo ==> of, for, to, do, go
/tmp/foo.txt:1445: Fo ==> Of, For, To, Do, Go
FIXED: /tmp/foo.txt

-------8<-------
SUMMARY:
ssee          2
wee           1
$ diff -u /tmp/foo.txt{,.orig}
--- /tmp/foo.txt	2022-11-17 17:05:57.752060107 +0100
+++ /tmp/foo.txt.orig	2022-11-17 17:03:26.323263279 +0100
@@ -1123,7 +1123,7 @@
  ("`si" ?ሢ) ("ssi" "ሢ") ("s2i" "ሢ")
  ("`sa" ?ሣ) ("ssa" "ሣ") ("s2a" "ሣ")
  ("`sE" ?ሤ) ("ssE" "ሤ") ("s2E" "ሤ")
-   ("`see" "ሤ") ("see" "ሤ") ("s2ee" "ሤ")
+   ("`see" "ሤ") ("ssee" "ሤ") ("s2ee" "ሤ")
  ("`s" ?ሥ) ("ss" "ሥ") ("s2" "ሥ")
  ("`so" ?ሦ) ("sso" "ሦ") ("s2o" "ሦ")
  ("`sWa" ?ሧ) ("ssWa" "ሧ") ("s2Wa" "ሧ")
@@ -1292,7 +1292,7 @@
  ("wu" ?ዉ)
  ("wi" ?ዊ)
  ("wa" ?ዋ)
- ("wE" ?ዌ) ("we" "ዌ")
+ ("wE" ?ዌ) ("wee" "ዌ")
  ("w" ?ው)
  ("wo" ?ዎ)
 
@@ -1420,7 +1420,7 @@
  ("`Si" ?ፂ) ("SSi" "ፂ") ("S2i" "ፂ")
  ("`Sa" ?ፃ) ("SSa" "ፃ") ("S2a" "ፃ")
  ("`SE" ?ፄ) ("SSE" "ፄ") ("S2E" "ፄ")
-   ("`See" "ፄ") ("see" "ፄ") ("S2ee" "ፄ")
+   ("`See" "ፄ") ("SSee" "ፄ") ("S2ee" "ፄ")
  ("`S" ?ፅ) ("SS" "ፅ") ("S2" "ፅ")
  ("`So" ?ፆ) ("SSo" "ፆ") ("S2o" "ፆ")
```

I expect the diff above to be empty in this case, as the lines are specified to be ignored. This works for any other lines I've tried it with.

Does that clarify the issue?

**Comment by DimitriPapadopoulos** [comment]
My wrong, sorry about that.

I tried to reproduce the initial case unsuccessfully, on an Ubuntu 18.04 machine:
```
$ cat > foo.txt
("`see" "ሤ") ("ssee" "ሤ") ("s2ee" "ሤ")
("wE" ?ዌ) ("wee" "ዌ")
("`See" "ፄ") ("SSee" "ፄ") ("S2ee" "ፄ")
$ 
$ codespell -x foo.txt foo.txt
$ 
$ codespell foo.txt
foo.txt:1: ssee ==> see
foo.txt:2: wee ==> we
foo.txt:3: SSee ==> see
$ 
```
Linux machines use UTF-8 locales nowadays. Do I need to run this on a different operating system with a non-UTF-8 locale, such as Windows?

**Comment by DimitriPapadopoulos** [comment]
Then I tried to reproduce your last example:
```
$ curl -s -o foo.txt https://git.savannah.gnu.org/gitweb/\?p\=emacs.git\;a\=blob_plain\;f\=lisp/language/ethio-util.el\;hb\=HEAD
$ 
$ file foo.txt 
foo.txt: Lisp/Scheme program, UTF-8 Unicode text
$ 
$ codespell -x foo.txt foo.txt
Traceback (most recent call last):
  File "/home/dp165978/.local/bin/codespell", line 8, in <module>
    sys.exit(_script_main())
  File "/volatile/src/codespell/codespell_lib/_codespell.py", line 814, in _script_main
    return main(*sys.argv[1:])
  File "/volatile/src/codespell/codespell_lib/_codespell.py", line 935, in main
    build_exclude_hashes(options.exclude_file, exclude_lines)
  File "/volatile/src/codespell/codespell_lib/_codespell.py", line 471, in build_exclude_hashes
    for line in f:
  File "/usr/lib/python3.8/codecs.py", line 322, in decode
    (result, consumed) = self._buffer_decode(data, self.errors, final)
UnicodeDecodeError: 'utf-8' codec can't decode byte 0xf6 in position 10: invalid start byte
$ 
$ codespell foo.txt
WARNING: Decoding file using encoding=utf-8 failed: foo.txt
WARNING: Trying next encoding iso-8859-1
foo.txt:1126: ssee ==> see
foo.txt:1186: bu ==> by, be, but, bug, bun, bud, buy, bum
foo.txt:1186: Bu ==> By, Be, But, Bug, Bun, Bud, Buy, Bum
foo.txt:1203: te ==> the, be, we, to
foo.txt:1207: tE ==> the, be, we, to
foo.txt:1295: wee ==> we
foo.txt:1381: Te ==> The, Be, We, To
foo.txt:1385: TE ==> THE, BE, WE, TO
foo.txt:1423: SSee ==> see
foo.txt:1433: fo ==> of, for, to, do, go
foo.txt:1433: Fo ==> Of, For, To, Do, Go
$ 
```
It looks like this file is not identified a valid UTF-8 file when reading it:
* when loading it as an exclude file, my (recent) version of codespell fails,
* when loading it as a file to check, codespell fails to read it as UTF-8 and switches to ISO 8859-1.

**Comment by skangas** [comment]
[Since you asked: I'm seeing this on Debian GNU/Linux.]

It turns out that this file is not actually valid UTF-8. The `file` command says:

```
ethio-util.el: Lisp/Scheme program, Non-ISO extended-ASCII text, with LF, NEL line terminators
```

It also turns out that we cannot really change this file to valid UTF-8, as it is used for conversion. See the [bug report here](http://debbugs.gnu.org/59341). The file displays correctly in Emacs, but other editors seem to struggle with it.

I'm not sure if this is a type of file that codespell would like to support. If not, I guess this can be closed as wontfix, and I will just have to ignore those files for spell checking purposes.

Thanks.

**Comment by DimitriPapadopoulos** [comment]
From [difference between utf-8-emacs and utf-8 on Emacs](https://stackoverflow.com/questions/18284637/difference-between-utf-8-emacs-and-utf-8-on-emacs#18286633):
> `utf-8-emacs` supports additional characters such as the internal representation of binary data. As this is a non-standard extension of Unicode, a separate encoding was defined for it so if you use `utf-8` you will not accidentally include these non-standard extensions which could confuse other software.
> 
> You can use either encoding for elisp, unless you need to include binary data or obscure characters that are not part of Unicode it won't make a difference.

> `utf-8-emacs` is the encoding used internally by Emacs. It's visible in a few places (e.g. auto-save files), but as a general rule you should never use it unless you know what you're doing.

**Comment by skangas** [comment]
This relates to converting from non-ISO (for example IS13194) character encodings to UTF-8. We cannot encode the characters as anything other than the actual codepoints themselves, as we need people with fonts that can display them to visually spot mistakes. Or that is my understanding.

And, yes, this file would combine "two distinct and incompatible standard encodings". Emacs can read such files just fine, as long as you use `utf-8-emacs` (which corresponds to the character representation Emacs uses internally). But it might be asking too much of other software to support this very special case, as I believe such files will be very rare outside of Emacs itself. Therefore, from the point of view of codespell, the best solution here might just be to ignore such files completely.

So I agree with your conclusions, and I believe this issue could be closed. Thanks for taking the time to look into it.

**Comment by DimitriPapadopoulos** [comment]
I am afraid most software associate a unique encoding with a text file. File that combine multiple encodings are usually seen as binary files, because reading them requires specific logic and dedicated code. Such "_binary_" files are probably outside the scope of codespell.
